cli/http_cache.rs

```diff
@@ -0,0 +1,184 @@
+// Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
+#![allow(unused)]
+
+use crate::fs as deno_fs;
+use deno_core::ErrBox;
+use std::collections::HashMap;
+use std::fs;
+use std::fs::File;
+use std::io::Read;
+use std::path::Path;
+use std::path::PathBuf;
+use url::Url;
+
+/// Turn base of url (scheme, hostname, port) into a valid filename.
+/// This method replaces port part with a special string token (because
+/// ":" cannot be used in filename on some platforms).
+pub fn base_url_to_filename(url: &Url) -> PathBuf {
+  let mut out = PathBuf::new();
+
+  let scheme = url.scheme();
+  out.push(scheme);
+
+  match scheme {
+    "http" | "https" => {
+      let host = url.host_str().unwrap();
+      let host_port = match url.port() {
+        Some(port) => format!("{}_PORT{}", host, port),
+        None => host.to_string(),
+      };
+      out.push(host_port);
+    }
+    scheme => {
+      unimplemented!(
+        "Don't know how to create cache name for scheme: {}",
+        scheme
+      );
+    }
+  };
+
+  out
+}
+
+/// Turn provided `url` into a hashed filename.
+/// URLs can contain a lot of characters that cannot be used
+/// in filenames (like "?", "#", ":"), so in order to cache
+/// them properly they are deterministically hashed into ASCII
+/// strings.
+fn url_to_filename(url: &Url) -> PathBuf {
+  let mut cache_filename = base_url_to_filename(url);
+
+  let mut rest_str = url.path().to_string();
+  if let Some(query) = url.query() {
+    rest_str.push_str("?");
+    rest_str.push_str(query);
+  }
+  // NOTE: fragment is omitted on purpose - it's not taken into
+  // account when caching - it denotes parts of webpage, which
+  // in case of static resources doesn't make much sense
+  let hashed_filename = crate::checksum::gen(vec![rest_str.as_bytes()]);
+  cache_filename.push(hashed_filename);
+  cache_filename
+}
+
+pub type HeadersMap = HashMap<String, String>;
```

Comment:
Headers are not unique key -> value pair. A better type is `Vec<(String, String)>`. The value might not be encodable into a string, so probably a still better type is `Vec<(String, Vec<u8>)>`.

Reply:
I casted them to Strings as we discussed - seems to work fine for now

Reply:
I don't care so much about the non-string header values. That's a very rare situation that is easy to fix when it does come up.

However, I do think you should be using `Vec<(String, String)>` because it's going to come up that multiple header keys can appear, and it's difficult to refactor code that assumes there is only one header value per key.

Reply:
That's already handled - headers with multiple values are joined and stored as a single string - probably not ideal solution but works for first pass

Reply:
If you insist.